File: src/VMS.Modules.Trips/Models/RouteModels.cs
namespace VMS.Modules.Trips.Models;

public sealed class RouteStopModel
{
    public long RouteStopId { get; set; }
    public int CityId { get; set; }
    public int Sequence { get; set; }
    public string StopType { get; set; } = string.Empty;
    public int? PlannedDurationMin { get; set; }
    public string? Remarks { get; set; }
}

public sealed class RouteModel
{
    public int RouteId { get; set; }
    public string RouteCode { get; set; } = string.Empty;
    public string RouteName { get; set; } = string.Empty;
    public int OriginCityId { get; set; }
    public int DestinationCityId { get; set; }
    public bool IsRoundTrip { get; set; }
    public decimal? DistanceKm { get; set; }
    public int? StandardDurationMin { get; set; }
    public string Status { get; set; } = string.Empty;
    public string? Remarks { get; set; }
    public IReadOnlyList<RouteStopModel> Stops { get; set; } = [];
}

public sealed class RouteStopInput
{
    public int CityId { get; set; }
    public string StopType { get; set; } = string.Empty;
    public int? PlannedDurationMin { get; set; }
    public string? Remarks { get; set; }
}

public sealed class CreateRouteRequest
{
    /// <summary>Auto-suggested (<c>RT-{OriginAbbr}-{DestAbbr}</c>, <c>-2</c> suffix if taken) when left blank.</summary>
    public string? RouteCode { get; set; }
    public string RouteName { get; set; } = string.Empty;
    public bool IsRoundTrip { get; set; }
    public decimal? DistanceKm { get; set; }
    public int? StandardDurationMin { get; set; }
    public string? Remarks { get; set; }
    public IReadOnlyList<RouteStopInput> Stops { get; set; } = [];
}

public sealed class UpdateRouteRequest
{
    public string RouteName { get; set; } = string.Empty;
    public bool IsRoundTrip { get; set; }
    public decimal? DistanceKm { get; set; }
    public int? StandardDurationMin { get; set; }
    public string? Status { get; set; }
    public string? Remarks { get; set; }
}

public sealed class UpdateRouteStopsRequest
{
    public IReadOnlyList<RouteStopInput> Stops { get; set; } = [];
}

File: VMSFrontend/src/app/core/route.models.ts
/** Routes and route stops (FSD §17, §48.3 screen 9). */
export type RouteStopType = 'Origin' | 'Pickup' | 'Via' | 'Delivery' | 'Destination';

export interface RouteStopModel {
  routeStopId: number;
  cityId: number;
  sequence: number;
  stopType: RouteStopType;
  plannedDurationMin?: number | null;
  remarks?: string | null;
}

export interface RouteModel {
  routeId: number;
  routeCode: string;
  routeName: string;
  originCityId: number;
  destinationCityId: number;
  isRoundTrip: boolean;
  distanceKm?: number | null;
  standardDurationMin?: number | null;
  status: 'Active' | 'Inactive';
  remarks?: string | null;
  stops: RouteStopModel[];
}

export interface RouteStopInput {
  cityId: number;
  stopType: RouteStopType;
  plannedDurationMin?: number | null;
  remarks?: string | null;
}

/** Left blank, the code is auto-suggested as `RT-{OriginAbbr}-{DestAbbr}` (`-2`, `-3`, … if taken). */
export interface CreateRouteRequest {
  routeCode?: string | null;
  routeName: string;
  isRoundTrip: boolean;
  distanceKm?: number | null;
  standardDurationMin?: number | null;
  remarks?: string | null;
  stops: RouteStopInput[];
}

export interface UpdateRouteRequest {
  routeName: string;
  isRoundTrip: boolean;
  distanceKm?: number | null;
  standardDurationMin?: number | null;
  status?: 'Active' | 'Inactive' | null;
  remarks?: string | null;
}

export interface UpdateRouteStopsRequest {
  stops: RouteStopInput[];
}
